curvefs/src/mds/topology/topology_manager.cpp

```diff
@@ -61,14 +61,26 @@ void TopologyManager::RegistMetaServer(const MetaServerRegistRequest *request,
         });
     if (1 == list.size()) {
         // report duplicated register (already a metaserver with same ip and
-        // port in the cluster) to promise the idempotence of the interface
-        MetaServer ms;
-        topology_->GetMetaServer(list[0], &ms);
-        response->set_statuscode(TopoStatusCode::TOPO_OK);
-        response->set_metaserverid(ms.GetId());
-        response->set_token(ms.GetToken());
-        LOG(WARNING) << "Received duplicated registMetaServer message, "
-                     << "hostip = " << hostIp << ", port = " << port;
+        // port in the cluster) to promise the idempotence of the interface.
+        // If metaserver has copyset, return TOPO_METASERVER_EXIST;
+        // else return OK
+        auto copysetList = topology_->GetCopySetsInMetaServer(list[0]);
+        if (copysetList.empty()) {
+            MetaServer ms;
+            topology_->GetMetaServer(list[0], &ms);
+            response->set_statuscode(TopoStatusCode::TOPO_OK);
+            response->set_metaserverid(ms.GetId());
+            response->set_token(ms.GetToken());
+            LOG(WARNING) << "Received duplicated registMetaServer message, "
+                         << "metaserver is empty, hostip = "
+                         << hostIp << ", port = " << port;
+        } else {
+            response->set_statuscode(TopoStatusCode::TOPO_METASERVER_EXIST);
+            LOG(ERROR) << "Received duplicated registMetaServer message, "
+                       << "metaserver is not empty, hostip = "
+                       << hostIp << ", port = " << port;
+        }
+
```

Comment:
In the scenario of disk crash, a metaserver crash after fix disk and start it can not register?

Reply:
If metaserver data exist, it will not need register.
If metaserver data not exist, it can use new port regist, or wait utill data recorve to other metaserver.

Reply:
> If metaserver data exist, it will not need register. If metaserver data not exist, it can use new port regist, or wait utill data recorve to other metaserver.

If so, the pr can not resolve the problem of related issue. And what‘s the point of  adding a meta file?

Reply:
If one disk is error, the copyset on it must schedule to other metaserver.
A new metaserver with same ip and port can be added to the cluster only when all copies on the metaserver are migrated.
如果metaserver上面的copyset没有都迁移走，新的metaserver加进去，这样mds没办法感知copyset其实已经降级了。所以在metaserver数据迁走之前，先阻止相同ip和port的metaserver加入。